tiledb/sm/cpp_api/arrow_io_impl.h

```diff
@@ -0,0 +1,834 @@
+    retval.type = dim.type();
+    retval.elem_size = tiledb::impl::type_size(dim.type());
+    retval.cell_val_num = dim.cell_val_num();
+    retval.arrow_large = false;
+    return retval;
+
+  } else {
+    throw TDB_LERROR("Schema does not have attribute named '" + name + "'");
+  }
+}
+
+/* ****************************** */
+/*        Helper functions        */
+/* ****************************** */
+
+// Reorder offsets into Arrow-compatible format
+// NOTE: this currently hard-codes an inplace conversion from
+//       uint64 (TileDB) to int32 ("arrow small") only
+void offsets_to_arrow(BufferInfo binfo) {
+  size_t elem_size = binfo.elem_size;
+
+  uint64_t* offsets = binfo.offsets;
+  int32_t* offsets_i32 = (int32_t*)binfo.offsets;
+
+  uint64_t idx = 1;  // important: don't divide by 0
+  while (idx < binfo.offset_num) {
+    // we must check for zeros here to handle a run of empty
+    // values (zero-length) at the start of the buffer
+    // note that given the i32 conversion, we can simply
+    // skip the assignment because
+    //   offsets[idx] == 0 => offsets_i32[i:i+1] == 0
+    if (offsets[idx] != 0)
+      offsets_i32[idx] = (int32_t)(offsets[idx] / elem_size);
+    ++idx;
+  }
+  offsets_i32[idx] = binfo.elem_num;
+}
+
+void check_arrow_schema(ArrowSchema* arw_schema) {
+  if (arw_schema == nullptr)
+    TDB_LERROR("[ArrowIO]: Invalid ArrowSchema object!");
+
+  // sanity check the arrow schema
+  if (arw_schema->release == nullptr)
+    TDB_LERROR(
+        "[ArrowIO]: Invalid ArrowSchema: cannot import released schema.");
+  if (arw_schema->format != std::string("+s"))
+    TDB_LERROR("[ArrowIO]: Unsupported ArrowSchema: must be struct (+s).");
+  if (arw_schema->n_children < 1)
+    TDB_LERROR("[ArrowIO]: Unsupported ArrowSchema with 0 children.");
+  if (arw_schema->children == nullptr)
+    TDB_LERROR(
+        "[ArrowIO]: Invalid ArrowSchema with n_children>0 and children==NULL");
+}
+
+/* ****************************** */
+/*  Arrow C API Struct wrappers   */
+/* ****************************** */
+
+// NOTE: These structs manage the lifetime of the contained C structs.
+// CAUTION: they do *not* manage the lifetime of the underlying buffers.
+
+struct CPPArrowSchema {
+  /*
+   * Initialize a CPPArrowSchema object
+   *
+   * The lifetime of this object is controlled by the
+   * release callback set in the ArrowSchema.
+   *
+   * Note that an ArrowSchema is *movable*, provided
+   * the release callback of the source is set to null.
+   */
+  CPPArrowSchema(
+      std::string name,
+      std::string format,
+      std::string metadata,
+      int64_t flags,
+      std::vector<ArrowSchema*> children,
+      std::shared_ptr<CPPArrowSchema> dictionary)
+      : format_(format)
+      , name_(name)
+      , metadata_(metadata)
+      , children_(children)
+      , dictionary_(dictionary) {
+    flags_ = flags;
+    n_children_ = children.size();
+
+    schema_ = (ArrowSchema*)malloc(sizeof(ArrowSchema));
```

Comment:
We should use `const_cast` and `static_cast` instead of c casting here and below.

Reply:
✔️